### PR TITLE
crimson: Add support for bench osd command

### PR DESCRIPTION
**Pull request by NitzanMordhai**
this commit adds support for the 'bench' admin command in the OSD,
allowing administrators to perform benchmark tests on the OSD. The
'bench' command accepts 4 optional parameters with the following
default values:

1. count - Total number of bytes to write (default: 1GB).
2. size - Block size for each write operation (default: 4MB).
3. object_size - Size of each object to write (default: 0).
4. object_num - Number of objects to write (default: 0).

The results of the benchmark are returned in a JSON formatted output,
which includes the following fields:

1. bytes_written - Total number of bytes written during the benchmark.
2. blocksize - Block size used for each write operation.
3. elapsed_sec - Total time taken to complete the benchmark in seconds.
4. bytes_per_sec - Write throughput in bytes per second.
5. iops - Number of input/output operations per second.

Example JSON output:

```json
{
  "osd_bench_results": {
    "bytes_written": 1073741824,
    "blocksize": 4194304,
    "elapsed_sec": 0.5,
    "bytes_per_sec": 2147483648,
    "iops": 512
  }
}
```

Fixes: https://tracker.ceph.com/issues/66380
Signed-off-by: Nitzan Mordechai <nmordech@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/admin/osd_admin.cc [hunk 1]
```diff
@@ -93,6 +93,80 @@ class SendBeaconHook : public AdminSocketHook {
...
+              ceph::bufferlist&& input) const final
+  {
+    int64_t count = cmd_getval_or<int64_t>(cmdmap, "count", 1LL << 30);
+    int64_t bsize = cmd_getval_or<int64_t>(cmdmap, "size", 4LL << 20);
+    int64_t osize = cmd_getval_or<int64_t>(cmdmap, "object_size", 0);
+    int64_t onum = cmd_getval_or<int64_t>(cmdmap, "object_num", 0);
+    double elapsed = 0.0;
+    auto duration = local_conf()->osd_bench_duration;
+
+    if (bsize > local_conf()->osd_bench_max_block_size) {
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:126]: Let's also keep the original comments

File: src/crimson/admin/osd_admin.cc [hunk 2]
```diff
@@ -93,6 +93,80 @@ class SendBeaconHook : public AdminSocketHook {
...
+      if (count > max_count) {
+        logger().info("bench count {} > osd_bench_small_size_max_iops {}",
+                      count, max_count);
+        return seastar::make_ready_future<tell_result_t>(-EINVAL, "count too large");
+      }
+    } else {
+      int64_t max_count = local_conf()->osd_bench_large_size_max_throughput * duration;
+      if (count > max_count) {
+        logger().info("bench count {} > osd_bench_large_size_max_throughput {}, assuming {}/s,"
+                       " for {} seconds, can cause ill effects on osd.",
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:141]: `Please adjust 'osd_bench_small_size_max_iops' with a higher value if you wish to use a higher 'count'.`

File: src/crimson/admin/osd_admin.cc [hunk 3]
```diff
@@ -93,6 +93,80 @@ class SendBeaconHook : public AdminSocketHook {
...
+  {
+    int64_t count = cmd_getval_or<int64_t>(cmdmap, "count", 1LL << 30);
+    int64_t bsize = cmd_getval_or<int64_t>(cmdmap, "size", 4LL << 20);
+    int64_t osize = cmd_getval_or<int64_t>(cmdmap, "object_size", 0);
+    int64_t onum = cmd_getval_or<int64_t>(cmdmap, "object_num", 0);
+    double elapsed = 0.0;
+    auto duration = local_conf()->osd_bench_duration;
+
+    if (bsize > local_conf()->osd_bench_max_block_size) {
+      logger().info("bench block size {} > osd_bench_max_block_size {}",
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:127]: Can we keep the full log messages?

File: src/crimson/admin/osd_admin.cc [hunk 4]
```diff
@@ -93,6 +93,80 @@ class SendBeaconHook : public AdminSocketHook {
...
+      }
+    } else {
+      int64_t max_count = local_conf()->osd_bench_large_size_max_throughput * duration;
+      if (count > max_count) {
+        logger().info("bench count {} > osd_bench_large_size_max_throughput {}, assuming {}/s,"
+                       " for {} seconds, can cause ill effects on osd.",
+                      count, local_conf()->osd_bench_large_size_max_throughput, bsize, duration);
+        return seastar::make_ready_future<tell_result_t>(-EINVAL, "count too large");
+      }
+    }
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:145]:
```
  if (osize && bsize > osize) {
    bsize = osize;
  }
```

File: src/crimson/osd/osd.cc [hunk 5]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+      return 0;
+    });
+  }).handle_exception([&ss](std::exception_ptr eptr) {
+    try
+    {
+      std::rethrow_exception(eptr);
+    } 
+    catch (const std::exception& e)
+    {
+      ss << "Error: " << e.what();
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1486]: You can simply use `eptr` for error logging. See other instances of `handle_exception` in Crimson.

File: src/crimson/osd/osd.cc [hunk 6]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+      return -1;
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1487]: return type should be `seastar::future<int>`

File: src/test/pybind/test_rados.py [hunk 7]
```diff
@@ -1371,7 +1371,7 @@ def test_monmap_dump(self):
 
     @pytest.mark.bench
     def test_osd_bench(self):
-        cmd = dict(prefix='bench', size=4096, count=8192)
+        cmd = dict(prefix='bench', size=4096, count=8192, object_size=2147483648, object_num=9999)
```
Comment (Matan-B) [inline src/test/pybind/test_rados.py:1374]: This is a common change, please move this into a separate commit in order to explain the rationale.

File: src/crimson/osd/osd.cc [hunk 8]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+      if (onum && osize) {
+        oid_str = fmt::format("disk_bw_test_{}", (int)(dis(gen) % onum));
+        offset = (dis(gen) % (osize / bsize)) * bsize;
+      } else {
+        oid_str = fmt::format("disk_bw_test_{}", i * bsize);
+      }
+      ghobject_t oid(hobject_t(sobject_t(object_t(oid_str), 0)),
+                                         ghobject_t::NO_GEN,
+                                         shard_id_t::NO_SHARD);
+      ceph::os::Transaction t;
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1461]: Each loop iteration creates a new transaction **without** submitting it.

File: src/crimson/osd/osd.cc [hunk 9]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+    std::string data(osize, 'a');
+    ceph::buffer::list bl(osize);
+    bl.append(data);
+
+    std::string oid_str = fmt::format("disk_bw_test_{}", i);
+    ghobject_t oid(hobject_t(sobject_t(object_t(oid_str), 0)),
+                                       ghobject_t::NO_GEN, 
+                                       shard_id_t::NO_SHARD);
+    ceph::os::Transaction t;
+    t.write(collection, oid, 0, data.size(), bl);
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1428]: ditto

**Review by Matan-B — CHANGES_REQUESTED** [review]
The transactions are not submitted to the store. Were you able to test this?

File: src/crimson/admin/osd_admin.cc [hunk 10]
```diff
@@ -93,6 +93,80 @@ class SendBeaconHook : public AdminSocketHook {
...
+      if (ret < 0) {
+        return seastar::make_ready_future<tell_result_t>(ret, "run_bench failed with error:" + ss.str());
+      }
+      unique_ptr<Formatter> f{Formatter::create(format, "json-pretty", "json-pretty")};
+      f->open_object_section("osd_bench_results");
+      f->dump_int("bytes_written", count);
+      f->dump_int("blocksize", bsize);
+      f->dump_float("elapsed_sec", elapsed);
+      f->dump_float("bytes_per_sec", (elapsed > 0) ? count / elapsed : 0);
+      f->dump_float("iops", (elapsed > 0) ? (count / elapsed) / bsize : 0);
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:159]: The JSON output is created from the parameters passed

File: src/crimson/osd/osd.cc [hunk 11]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+    }).then([=, this, &collection]() {
+      return store.get_sharded_store().open_collection(collection).then(
+        [] (auto collection_ref) {
+        return collection_ref->flush_commit().then([=] (bool) {
+          return seastar::make_ready_future<>();
+        });
+      });
+    }).then([&elapsed, start]() {
+      seastar::timer<>::time_point end = std::chrono::steady_clock::now();
+      *elapsed = std::chrono::duration<double>(end - start).count();
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1476]: We should wait for the transactions to take place for measuring the elapsed time

File: src/test/pybind/test_rados.py (hunk 7 above)
Comment (NitzanMordhai) [reply src/test/pybind/test_rados.py:1374]: it shouldn't get into that PR at all, removing it (it was for debugging only)

File: src/crimson/admin/osd_admin.cc (hunk 10 above)
Comment (NitzanMordhai) [reply src/crimson/admin/osd_admin.cc:168]: it is using it (format) and create with it formatter

File: src/crimson/osd/osd.cc [hunk 12]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+        return seastar::make_ready_future<>();
+      }).then([this, &t, &cleanup]() {
+        return store.get_sharded_store().open_collection(coll_t::meta()).then(
+          [this, &t, &cleanup](auto collection_ref) {
+            return store.get_sharded_store().do_transaction(
+              collection_ref, std::move(t)).then([this, &cleanup, collection_ref] {
+                return store.get_sharded_store().do_transaction(
+                  collection_ref, std::move(cleanup)).then([collection_ref] {
+                    return collection_ref->flush_commit().then([](bool) {
+                      return seastar::make_ready_future<>();
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1433]: Why is this needed?

File: src/crimson/osd/osd.cc [hunk 13]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+        std::string oid_str = fmt::format("disk_bw_test_{}", i);
+        ghobject_t oid(hobject_t(sobject_t(object_t(oid_str), 0)),
+          ghobject_t::NO_GEN,
+          shard_id_t::NO_SHARD);
+        t.write(coll_t::meta(), oid, 0, data.size(), bl);
+        cleanup.remove(coll_t::meta(), oid);
+        return seastar::make_ready_future<>();
+      }).then([this, &t, &cleanup]() {
+        return store.get_sharded_store().open_collection(coll_t::meta()).then(
+          [this, &t, &cleanup](auto collection_ref) {
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1427]: I think that we can reuse `pg_shard_manager.get_meta_coll()` instead of creating a new one

File: src/crimson/osd/osd.cc [hunk 14]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+            return store.get_sharded_store().do_transaction(
+              collection_ref, std::move(t)).then([this, &cleanup, collection_ref] {
+                return store.get_sharded_store().do_transaction(
+                  collection_ref, std::move(cleanup)).then([collection_ref] {
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1431]: Cleanup phase should not be part of the test results (elapsed time), we probably shouldn't submit `cleanup` here.

File: src/crimson/osd/osd.cc [hunk 15]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+                      });
+                  });
+          }).then([start, elapsed]() {
+            seastar::timer<>::time_point end = std::chrono::steady_clock::now();
+            *elapsed = std::chrono::duration<double>(end - start).count();
+            return seastar::make_ready_future<>();
+          }).handle_exception([FNAME, elapsed](auto eptr) {
+            ERROR("error in run_bench: {}", eptr);
+            *elapsed = -1;
+            return seastar::make_ready_future<>();
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1486]: `elapsed` should be carried by the returned future - `seastar::make_ready_future<double>(elapsed)`

File: src/crimson/osd/osd.cc [hunk 16]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+        std::string data(osize, 'a');
+        ceph::buffer::list bl(osize);
+        bl.append(data);
+        std::string oid_str = fmt::format("disk_bw_test_{}", i);
+        ghobject_t oid(hobject_t(sobject_t(object_t(oid_str), 0)),
+          ghobject_t::NO_GEN,
+          shard_id_t::NO_SHARD);
+        t.write(coll_t::meta(), oid, 0, data.size(), bl);
+        cleanup.remove(coll_t::meta(), oid);
+        return seastar::make_ready_future<>();
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1424]: ~In order to simplify this function, the transaction creation logic can be voided (aka regular for loop). Only the submission itself should be futurized.~
However, when looking again, it might be more correct and correlated with Classic's bench test if we'll submit each write with a separate transaction - and later on we could await all of them using `seastar::when_all_succeed`.
For cleanup we can use a single transaction (as used here).

File: src/crimson/admin/osd_admin.cc [hunk 17]
```diff
@@ -93,6 +93,89 @@ class SendBeaconHook : public AdminSocketHook {
...
+      // having a sane value.  If we allow any block size to be set things
+      // can still go sideways.
+      logger().info("block 'size' values are capped at {}. If you wish to use"
+                    " a higher value, please adjust 'osd_bench_max_block_size'",
+                    byte_u_t(local_conf()->osd_bench_max_block_size));
+      return seastar::make_ready_future<tell_result_t>(-EINVAL, "block size too large");
+    } else if (bsize < (1LL << 20)) {
+      int64_t max_count = bsize * duration * local_conf()->osd_bench_small_size_max_iops;
+      if (count > max_count) {
+        logger().info("bench count {} > osd_bench_small_size_max_iops {}",
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:137]: See `crimson/common/log.h`, can you please use `INFO` and `DEBUG` macros?

File: src/include/types.h [hunk 18]
```diff
@@ -482,6 +482,11 @@ struct byte_u_t {
   explicit byte_u_t(uint64_t _v) : v(_v) {};
 };
 
+namespace fmt {
+  template <>
+  struct formatter<byte_u_t> : ostream_formatter {};
+}
+
```
Comment (Matan-B) [inline src/include/types.h:489]:
```
#if FMT_VERSION >= 90000
template <> struct fmt::formatter<byte_u_t> : fmt::ostream_formatter {};
#endif
```

File: src/crimson/osd/osd.cc [hunk 19]
```diff
@@ -1402,6 +1403,93 @@ seastar::future<> OSD::send_beacon()
...
+                    return store.get_sharded_store().do_transaction(
+                      collection_ref, std::move(t)).then([this, &cleanup, collection_ref] {
+                        return store.get_sharded_store().do_transaction(
+                          collection_ref, std::move(cleanup)).then([collection_ref] {
+                            return collection_ref->flush_commit().then([](bool) {
+                              return seastar::make_ready_future<>();
+                            });
+                          });
+                      });
+                  });
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1478]: Same as previous comments.

File: src/crimson/admin/osd_admin.cc (hunk 10 above)
Comment (Matan-B) [reply src/crimson/admin/osd_admin.cc:168]: I meant that this are not the actual results (expect from `elapsed`) in order to explain how the previous version of this PR was able to output results.

File: src/crimson/osd/osd.cc (hunk 12 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1433]: i followed the original run, and in case we have some flush_commit in the future, but for now, it is not necessary, do you think we will have

File: src/crimson/osd/osd.cc (hunk 13 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1427]: i think that open_collection finds existing collections (by cid) before it tries to create a new one..

File: src/crimson/admin/osd_admin.cc (hunk 17 above)
Comment (NitzanMordhai) [reply src/crimson/admin/osd_admin.cc:137]: i don't mind to use INFO or DEBUG, but all the other using logger() in osd_admin , will change that

File: src/crimson/osd/osd.cc (hunk 19 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1478]: classic check elapsed for the second loop only, which means we can leave the previous loop as it is and change that one only.

File: src/crimson/osd/osd.cc (hunk 12 above)
Comment (Matan-B) [reply src/crimson/osd/osd.cc:1435]: Yes, `FuturizedCollection::flush_commit` is not implemented in Crimson.
Classic uses it in order to await all the transactions to be committed:
```
    /**
     * Async flush_commit
     *
     * There are two cases:
     * 1) collection is currently idle: the method returns true.  c is
     *    not touched.
     * 2) collection is not idle: the method returns false and c is
     *    called asynchronously with a value of 0 once all transactions
     *    queued on this collection prior to the call have been applied
     *    and committed.
     */
```

File: src/crimson/osd/osd.cc (hunk 13 above)
Comment (Matan-B) [reply src/crimson/osd/osd.cc:1429]: This will work. You can regard this suggestion as a nit, since `get_meta_coll` is already implemented.

File: src/crimson/admin/osd_admin.cc (hunk 17 above)
Comment (Matan-B) [reply src/crimson/admin/osd_admin.cc]: These macros were introduced recently, we are moving gradually :)

File: src/crimson/osd/osd.cc (hunk 19 above)
Comment (Matan-B) [reply src/crimson/osd/osd.cc]: I was referring to the `flush_commit` and `open_collection` usage comments.

File: src/crimson/osd/osd.cc [hunk 20]
```diff
@@ -1402,6 +1403,83 @@ seastar::future<> OSD::send_beacon()
...
+          collection_ref, std::move(cleanup_t)));
+      }
+    }
+    
+    auto start = std::chrono::steady_clock::now();
+    co_await seastar::when_all_succeed(futures_bench.begin(), futures_bench.end());
+    auto end = std::chrono::steady_clock::now();
+    double elapsed = std::chrono::duration<double>(end - start).count();
+    co_await seastar::when_all_succeed(cleanup_futures.begin(), cleanup_futures.end());
+    co_return co_await seastar::make_ready_future<double>(elapsed);
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1480]: :+1:

File: src/crimson/admin/osd_admin.cc [hunk 21]
```diff
@@ -93,6 +94,91 @@ class SendBeaconHook : public AdminSocketHook {
...
+                      duration);
+        return seastar::make_ready_future<tell_result_t>(-EINVAL, "count too large");
+      }
+    }
+    if (osize && bsize > osize) {
+      bsize = osize;
+    }
+
+    return osd.run_bench(count, bsize, osize, onum).then(
+      [=](double elapsed) {
```
Comment (Matan-B) [inline src/crimson/admin/osd_admin.cc:159]: capture clause should be empty

**Review by Matan-B — COMMENTED** [review]
LGTM, great co routines usage!

File: src/crimson/osd/osd.cc [hunk 22]
```diff
@@ -1402,6 +1403,82 @@ seastar::future<> OSD::send_beacon()
...
+      bl.push_back(std::move(bp));
+      bl.rebuild_page_aligned();
+
+      std::string oid_str;
+      uint64_t offset = 0;
+      if (onum && osize) {
+          oid_str = fmt::format("disk_bw_test_{}", dis(gen) % onum);
+          offset = (dis(gen) % (osize / bsize)) * bsize;
+      } else {
+          oid_str = fmt::format("disk_bw_test_{}", i * bsize);
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1458]: indentation

File: src/crimson/osd/osd.cc [hunk 23]
```diff
@@ -1402,6 +1403,82 @@ seastar::future<> OSD::send_beacon()
   return monc->send_message(std::move(beacon));
 }
 
+seastar::future<double> OSD::run_bench(int64_t count, int64_t bsize, int64_t osize, int64_t onum) {
+    LOG_PREFIX(OSD::run_bench);
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1407]: make check failed with:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/osd.cc:1407:5: error: unused variable 'FNAME' [-Werror,-Wunused-variable]
    LOG_PREFIX(OSD::run_bench);
```
You can add a `DEBUG();` a line after that to fix the error

**Review by Matan-B — COMMENTED** [review]
lgtm, final comments

File: src/crimson/osd/osd.cc [hunk 24]
```diff
@@ -1402,6 +1403,82 @@ seastar::future<> OSD::send_beacon()
...
+
+      if (!onum || !osize) {
+        cleanup_t.remove(coll_t::meta(), oid);
+        cleanup_futures.push_back(store.get_sharded_store().do_transaction(
+          collection_ref, std::move(cleanup_t)));
+      }
+    }
+    
+    auto start = std::chrono::steady_clock::now();
+    co_await seastar::when_all_succeed(futures_bench.begin(), futures_bench.end());
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:1475]: This may not be accurate. I was wrong here earlier since once we push a future to `futures_bench` it will actually get executed regardless if we wait for it later on (via `when_all_succeed`) or not. We should probably move `start` higher in the flow.

File: src/crimson/osd/osd.cc (hunk 22 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1458]: done

File: src/crimson/admin/osd_admin.cc [hunk 25]
```diff
@@ -93,6 +94,104 @@ class SendBeaconHook : public AdminSocketHook {
...
+ */
+class RunOSDBenchHook : public AdminSocketHook {
+public:
+  explicit RunOSDBenchHook(crimson::osd::OSD& osd) :
+    AdminSocketHook{"bench",
+        "name=count,type=CephInt,req=false "
+        "name=size,type=CephInt,req=false "
+        "name=object_size,type=CephInt,req=false "
+        "name=object_num,type=CephInt,req=false",
+        "run OSD bench"},
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/crimson/admin/osd_admin.cc:113]: done

File: src/crimson/admin/osd_admin.cc [hunk 26]
```diff
@@ -93,6 +94,104 @@ class SendBeaconHook : public AdminSocketHook {
...
+      }
+    }
+    if (osize && bsize > osize) {
+      bsize = osize;
+    }
+
+    return osd.run_bench(count, bsize, osize, onum).then(
+      [format, bsize, count](double elapsed) {
+      if (elapsed < 0) {
+        return seastar::make_ready_future<tell_result_t>(elapsed, "bench failed with error");
```
[the remark this answers is not in the document]
Comment (NitzanMordhai) [reply src/crimson/admin/osd_admin.cc:174]: done

File: src/crimson/osd/osd.cc (hunk 23 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1407]: done

File: src/crimson/osd/osd.cc (hunk 24 above)
Comment (NitzanMordhai) [reply src/crimson/osd/osd.cc:1475]: done

File: src/crimson/osd/osd.h [hunk 27]
```diff
@@ -247,6 +247,10 @@ class OSD final : public crimson::net::Dispatcher,
 
 public:
   seastar::future<> send_beacon();
+  seastar::future<double> run_bench(int64_t count,
+  int64_t bsize,
+  int64_t osize,
+  int64_t onum);
```
Comment (Matan-B) [inline src/crimson/osd/osd.h:253]: nit: indentation :)

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
@NitzanMordhai, can you please run this through the suite?

**Comment by NitzanMordhai** [comment]
> @NitzanMordhai, can you please run this through the suite?

running: https://pulpito.ceph.com/nmordech-2024-08-15_10:15:18-crimson-rados-wip-nitzan-bench-osd-admin-command-distro-default-smithi/

**Comment by NitzanMordhai** [comment]
@Matan-B all check pass, there are 2 unrelated errors

**Comment by Matan-B** [comment]
> @Matan-B all check pass, there are 2 unrelated errors

Thanks! Can you please reference here the trackers of the errors? Should help with understanding the suite's status. 

Should the last commit be removed? 72dab7e370d3d3835fed2bb6aa7ec7741f8ce33f

**Comment by Matan-B** [comment]
@NitzanMordhai

**Comment by NitzanMordhai** [comment]
> Should the last commit be removed? [72dab7e](https://github.com/ceph/ceph/commit/72dab7e370d3d3835fed2bb6aa7ec7741f8ce33f)

we can remove and add it back in case we need to debug something

**Comment by NitzanMordhai** [comment]
unrelated failures: https://tracker.ceph.com/issues/67568

**Comment by Matan-B** [comment]
```
2024-08-15T10:56:39.395 INFO:tasks.workunit.client.0.smithi097.stdout:../../../clone.client.0/src/test/pybind/test_rados.py::TestCommand::test_osd_bench PASSED [ 96%]
```
